File: backend/tools/actions.js
import { readJson, sendJson } from "../shared/http.js";

export const actionsTool = {
  id: "actions",
  name: "Actions",
  capabilities: [
    {
      id: "actions.simulate",
      name: "Simulate Action",
      description: "Pretend to execute a tool call and return a traceable result.",
      method: "POST",
      endpoint: "/actions/simulate"
    }
  ],
  async route({ request, response, url }) {
    if (request.method !== "POST" || url.pathname !== "/actions/simulate") {
      return false;
    }

    const payload = await readJson(request);
    sendJson(response, 200, {
      ok: true,
      actionId: `act_${Date.now()}`,
      capability: payload.capability || "unknown",
      status: "simulated",
      received: payload
    });
    return true;
  }
};

File: backend/app/tools/actions.py
from time import time
from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel, Field

tool_name = "actions"
router = APIRouter(prefix=f"/{tool_name}", tags=[tool_name])


class SimulateActionRequest(BaseModel):
    capability: str = "unknown"
    input: dict[str, Any] = Field(default_factory=dict)


class SimulateActionResponse(BaseModel):
    ok: bool
    action_id: str
    capability: str
    status: str
    received: SimulateActionRequest


capabilities = [
    {
        "id": "actions.simulate",
        "tool": "actions",
        "name": "Simulate Action",
        "description": "Pretend to execute a tool call and return a traceable result.",
        "method": "POST",
        "endpoint": "/actions/simulate",
        "input_model": "SimulateActionRequest",
        "output_model": "SimulateActionResponse",
        "tags": ["debug", "action"],
    }
]


@router.post("/simulate", response_model=SimulateActionResponse)
def simulate_action(payload: SimulateActionRequest):
    return SimulateActionResponse(
        ok=True,
        action_id=f"act_{int(time() * 1000)}",
        capability=payload.capability,
        status="simulated",
        received=payload,
    )
